service/handlers.go

```diff
@@ -0,0 +1,207 @@
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+
+	c.JSON(http.StatusOK, map[string]string{
+		"response": "Book inserted at ID: " + id,
+	})
+}
+
+func UpdateBook(c *gin.Context) {
+	updateUserData(c)
+	db := data_store.NewBookStorer()
+	if db == nil {
+		c.JSON(http.StatusInternalServerError, INIT_BOOKSTORE_FAIL)
+		return
+	}
+
+	id := c.Param("id")
+	var book models.Book
+	err := c.ShouldBind(&book)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+
+	err = db.UpdateBook(book.Title, id)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+
+	c.JSON(http.StatusOK, map[string]string{
+		"response": "Book updated at ID: " + id,
+	})
+}
+
+func DeleteBook(c *gin.Context) {
+	updateUserData(c)
+	db := data_store.NewBookStorer()
+	if db == nil {
+		c.JSON(http.StatusInternalServerError, INIT_BOOKSTORE_FAIL)
+		return
+	}
+
+	id := c.Param("id")
+	err := db.DeleteBook(id)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+
+	c.JSON(http.StatusOK, map[string]string{
+		"response": "Book deleted at ID: " + id,
+	})
+}
+
+func SearchBook(c *gin.Context) {
+	updateUserData(c)
+	db := data_store.NewBookStorer()
+	if db == nil {
+		c.JSON(http.StatusInternalServerError, INIT_BOOKSTORE_FAIL)
+		return
+	}
+
+	title := c.DefaultQuery("title", "")
+	authorName := c.DefaultQuery("author_name", "")
+	priceRangeStr := c.DefaultQuery("price_range", "")
+
+	searchResult, err := db.SearchBook(title, authorName, priceRangeStr)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+
+	if searchResult == nil {
+		c.JSON(http.StatusInternalServerError, INALID_INPUT_FAIL)
+		return
+	}
+
+	c.JSON(http.StatusOK, searchResult)
+}
+
+func GetStoreInfo(c *gin.Context) {
+	updateUserData(c)
+	db := data_store.NewBookStorer()
+	if db == nil {
+		c.JSON(http.StatusInternalServerError, INIT_BOOKSTORE_FAIL)
+		return
+	}
+
+	count, diffAuthors, err := db.GetStoreInfo()
+	if diffAuthors == nil {
+		c.JSON(http.StatusInternalServerError, STORE_INFO_FAIL)
+	}
+
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err)
+		return
+	}
+
+	response := fmt.Sprintf("There are %d books by %.0f different authors", count, *diffAuthors)
+	c.JSON(http.StatusOK, map[string]string{
+		"response": response,
+	})
+
+}
+
+func GetUserData(c *gin.Context) {
+	username := c.DefaultQuery("username", "")
+	if username == "" {
+		return
+	}
+
+	r := data_store.NewUserDater()
+	if r == nil {
+		c.JSON(http.StatusInternalServerError, INIT_USERDATA_FAIL)
+	}
+
+	response := r.GetUserActivity(username)
+	if response == nil {
+		return
+	}
+
+	userPath1 := fmt.Sprintf("%v", response[0])
+	userPath2 := fmt.Sprintf("%v", response[1])
+	userPath3 := fmt.Sprintf("%v", response[2])
+
+	userData := map[string]string{"path1": userPath1}
+	if userPath2 != "" {
+		userData["path2"] = userPath2
+		if userPath3 != "" {
+			userData["path3"] = userPath3
+		}
+	}
```

Comment:
use should return an array from `GetUserActivity`
Why you're returning the 'Path' prefix to a client? it is for your internal implemention...

Reply:
_"use should return an array from GetUserActivity"_ - just an array?  I recall we talked about always returning a JSON. Is it not the case for user activety? do you mean an array of JSON's? please let me know.

_"Why you're returning the Path prefix to the client? it is for your internal implementation"_ - I want to let the client know which of the three returned paths is the most and least recent. Definitely  can use better names, I will change.